### PR TITLE
universal bootstrap

### PR DESCRIPTION
**Pull request by weissi**
This is the universal bootstrap work captured in the following four PRs:

- https://github.com/apple/swift-nio/pull/1323
- https://github.com/apple/swift-nio-transport-services/pull/69
- https://github.com/apple/swift-nio-ssl/pull/185
- https://github.com/apple/swift-nio-examples/pull/47

### CODE REVIEW

File: Sources/NIOTransportServices/NIOTSEventLoopGroup.swift [hunk 1]
```diff
@@ -85,4 +85,16 @@ public final class NIOTSEventLoopGroup: EventLoopGroup {
         return EventLoopIterator(self.eventLoops)
     }
 }
+
+@available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *)
+public struct NIOTSClientTLSProvider: NIOClientTLSProvider {
+    public typealias Bootstrap = NIOTSConnectionBootstrap
+
+    public init() {
```
Comment (glbrntt) [inline Sources/NIOTransportServices/NIOTSEventLoopGroup.swift]: Shouldn't the user be able to specify the `tlsOptions` by providing them here?

File: Sources/NIOTransportServices/NIOTSEventLoopGroup.swift [hunk 2]
```diff
@@ -85,4 +85,16 @@ public final class NIOTSEventLoopGroup: EventLoopGroup {
         return EventLoopIterator(self.eventLoops)
     }
 }
+
+@available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *)
+public struct NIOTSClientTLSProvider: NIOClientTLSProvider {
```
Comment (Lukasa) [inline Sources/NIOTransportServices/NIOTSEventLoopGroup.swift]: Can we add a doc comment here?

Comment (weissi) [reply Sources/NIOTransportServices/NIOTSEventLoopGroup.swift]: thanks, done

File: Sources/NIOTransportServices/NIOTSEventLoopGroup.swift (hunk 1 above)
Comment (weissi) [reply Sources/NIOTransportServices/NIOTSEventLoopGroup.swift]: @glbrntt thank you, yes!

File: Sources/NIOTransportServices/NIOTSEventLoopGroup.swift [hunk 3]
```diff
@@ -85,4 +85,34 @@ public final class NIOTSEventLoopGroup: EventLoopGroup {
...
+///
+/// Example:
+///
+///     // Creating the "universal bootstrap" with the `NIOTSClientTLSProvider`.
+///     let tlsProvider = NIOTSClientTLSProvider()
+///     let bootstrap = NIOClientTCPBootstrap(NIOTSConnectionBootstrap(group: group), tls: tlsProvider)
+///
+///     // Bootstrapping a connection using the "universal bootstrapping mechanism"
+///     let connection = bootstrap.enableTLS()
+///                          .connect(to: "example.com")
```
Comment (glbrntt) [inline Sources/NIOTransportServices/NIOTSEventLoopGroup.swift]: Does `connect(to: String)` exist?!

Comment (weissi) [reply Sources/NIOTransportServices/NIOTSEventLoopGroup.swift]: whoops, no. `host:port:`. great catch, thanks, fixed

**Comment by weissi** [comment]
Thanks @glbrntt I think I addressed it all

**Pull request merged**